File: src/environment.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const types = @import("types.zig");
const Variable = types.Variable;

const EnvironmentConfig = struct {
    environments: std.StringHashMap(std.StringHashMap([]const u8)),

    pub fn init(allocator: Allocator) EnvironmentConfig {
        return EnvironmentConfig{
            .environments = std.StringHashMap(std.StringHashMap([]const u8)).init(allocator),
        };
    }

    pub fn deinit(self: *EnvironmentConfig, allocator: Allocator) void {
        var env_iterator = self.environments.iterator();
        while (env_iterator.next()) |env_entry| {
            var var_iterator = env_entry.value_ptr.iterator();
            while (var_iterator.next()) |var_entry| {
                allocator.free(var_entry.key_ptr.*);
                allocator.free(var_entry.value_ptr.*);
            }
            env_entry.value_ptr.deinit();
            allocator.free(env_entry.key_ptr.*);
        }
        self.environments.deinit();
    }
};

pub fn loadEnvironmentFile(allocator: Allocator, http_file_path: []const u8, environment_name: ?[]const u8) !std.ArrayList(Variable) {
    var variables = std.ArrayList(Variable).initCapacity(allocator, 0) catch @panic("OOM");

    if (environment_name == null) {
        return variables;
    }

    const env_file_path = try findEnvironmentFile(allocator, http_file_path);
    defer if (env_file_path) |path| allocator.free(path);

    if (env_file_path == null) {
        return variables;
    }

    const env_config = parseEnvironmentFile(allocator, env_file_path.?) catch {
        // If we can't parse the environment file, just return empty variables
        // This allows the tool to continue working even if the env file has issues
        return variables;
    };
    defer {
        var config = env_config;
        config.deinit(allocator);
    }

    if (env_config.environments.get(environment_name.?)) |env_vars| {
        var iterator = env_vars.iterator();
        while (iterator.next()) |entry| {
            try variables.append(allocator, .{
                .name = try allocator.dupe(u8, entry.key_ptr.*),
                .value = try allocator.dupe(u8, entry.value_ptr.*),
            });
        }
    }

    return variables;
}

fn findEnvironmentFile(allocator: Allocator, http_file_path: []const u8) !?[]const u8 {
    const dirname = std.fs.path.dirname(http_file_path) orelse ".";

    var current_dir = try allocator.dupe(u8, dirname);
    defer allocator.free(current_dir);

    while (true) {
        const env_file_path = try std.fs.path.join(allocator, &[_][]const u8{ current_dir, "http-client.env.json" });
        defer allocator.free(env_file_path);

        if (std.fs.cwd().access(env_file_path, .{})) |_| {
            return try allocator.dupe(u8, env_file_path);
        } else |_| {
            const parent_dir = std.fs.path.dirname(current_dir);
            if (parent_dir == null or std.mem.eql(u8, parent_dir.?, current_dir)) {
                break;
            }

            const new_current_dir = try allocator.dupe(u8, parent_dir.?);
            allocator.free(current_dir);
            current_dir = new_current_dir;
        }
    }

    return null;
}

fn parseEnvironmentFile(allocator: Allocator, file_path: []const u8) !EnvironmentConfig {
    const file = try std.fs.cwd().openFile(file_path, .{});
    defer file.close();

    const file_size = try file.getEndPos();
    const content = try allocator.alloc(u8, file_size);
    defer allocator.free(content);
    _ = try file.readAll(content);

    var config = EnvironmentConfig.init(allocator);

    const parsed = std.json.parseFromSlice(std.json.Value, allocator, content, .{}) catch |err| {
        return err;
    };
    defer parsed.deinit();

    const root = parsed.value;
    if (root != .object) {
        return error.InvalidEnvironmentFile;
    }

    var env_iterator = root.object.iterator();
    while (env_iterator.next()) |env_entry| {
        const env_name = try allocator.dupe(u8, env_entry.key_ptr.*);
        var env_vars = std.StringHashMap([]const u8).init(allocator);

        if (env_entry.value_ptr.* == .object) {
            var var_iterator = env_entry.value_ptr.object.iterator();
            while (var_iterator.next()) |var_entry| {
                const var_name = try allocator.dupe(u8, var_entry.key_ptr.*);
                const var_value = switch (var_entry.value_ptr.*) {
                    .string => |s| try allocator.dupe(u8, s),
                    else => try allocator.dupe(u8, ""),
                };
                try env_vars.put(var_name, var_value);
            }
        }

        try config.environments.put(env_name, env_vars);
    }

    return config;
}

File: src/environment.rs
use crate::types::Variable;
use anyhow::Result;
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};

pub fn load_environment_file(
    http_file_path: &str,
    environment_name: Option<&str>,
) -> Result<Vec<Variable>> {
    let mut variables = Vec::new();

    if environment_name.is_none() {
        return Ok(variables);
    }

    let env_file_path = find_environment_file(http_file_path)?;
    if env_file_path.is_none() {
        return Ok(variables);
    }

    let env_config = parse_environment_file(&env_file_path.unwrap())?;

    if let Some(env_vars) = env_config.get(environment_name.unwrap()) {
        for (name, value) in env_vars {
            variables.push(Variable {
                name: name.clone(),
                value: value.clone(),
            });
        }
    }

    Ok(variables)
}

fn find_environment_file(http_file_path: &str) -> Result<Option<PathBuf>> {
    let path = Path::new(http_file_path);
    let mut current_dir = path
        .parent()
        .unwrap_or_else(|| Path::new("."))
        .to_path_buf();

    loop {
        let env_file = current_dir.join("http-client.env.json");
        if env_file.exists() {
            return Ok(Some(env_file));
        }

        if let Some(parent) = current_dir.parent() {
            current_dir = parent.to_path_buf();
        } else {
            break;
        }
    }

    Ok(None)
}

fn parse_environment_file(file_path: &Path) -> Result<HashMap<String, HashMap<String, String>>> {
    let content = fs::read_to_string(file_path)?;
    let json: Value = serde_json::from_str(&content)?;

    let mut config = HashMap::new();

    if let Value::Object(root) = json {
        for (env_name, env_value) in root {
            let mut env_vars = HashMap::new();

            if let Value::Object(vars) = env_value {
                for (var_name, var_value) in vars {
                    let value_str = match var_value {
                        Value::String(s) => s.clone(),
                        _ => String::new(),
                    };
                    env_vars.insert(var_name.clone(), value_str);
                }
            }

            config.insert(env_name.clone(), env_vars);
        }
    }

    Ok(config)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;
    use tempfile::tempdir;

    #[test]
    fn load_environment_file_returns_empty_without_environment() {
        let temp = tempdir().unwrap();
        let http_file = temp.path().join("request.http");
        let vars = load_environment_file(http_file.to_str().unwrap(), None).unwrap();
        assert!(vars.is_empty());
    }

    #[test]
    fn load_environment_file_reads_nearest_env_file() {
        let temp = tempdir().unwrap();
        let nested = temp.path().join("nested");
        fs::create_dir(&nested).unwrap();

        let http_file = nested.join("request.http");
        fs::write(&http_file, "GET http://example.com").unwrap();
        fs::write(
            temp.path().join("http-client.env.json"),
            r#"{"dev":{"TOKEN":"abc","COUNT":1,"FLAG":true}}"#,
        )
        .unwrap();

        let vars = load_environment_file(http_file.to_str().unwrap(), Some("dev")).unwrap();
        let map: std::collections::HashMap<_, _> =
            vars.into_iter().map(|v| (v.name, v.value)).collect();
        assert_eq!(map.get("TOKEN"), Some(&"abc".to_string()));
        assert_eq!(map.get("COUNT"), Some(&String::new()));
        assert_eq!(map.get("FLAG"), Some(&String::new()));
    }

    #[test]
    fn find_environment_file_returns_none_when_absent() {
        let temp = tempdir().unwrap();
        let http_file = temp.path().join("request.http");
        fs::write(&http_file, "GET http://example.com").unwrap();
        let found = find_environment_file(http_file.to_str().unwrap()).unwrap();
        assert!(found.is_none());
    }

    #[test]
    fn parse_environment_file_handles_non_string_values() {
        let temp = tempdir().unwrap();
        let env_file = temp.path().join("http-client.env.json");
        fs::write(
            &env_file,
            r#"{"dev":{"TEXT":"value","NUMBER":123,"OBJECT":{"foo":"bar"}}}"#,
        )
        .unwrap();

        let parsed = parse_environment_file(&env_file).unwrap();
        let dev = parsed.get("dev").unwrap();
        assert_eq!(dev.get("TEXT").unwrap(), "value");
        assert_eq!(dev.get("NUMBER").unwrap(), "");
        assert_eq!(dev.get("OBJECT").unwrap(), "");
    }
}
